python/ray/_private/runtime_env/working_dir.py

```diff
@@ -317,128 +299,164 @@ def package_exists(pkg_uri: str) -> bool:
+
+        It'll check whether the runtime environment exists in the cluster or
+        not. If it doesn't, a package will be created based on the working
+        directory and modules defined in job config. The package will be
+        uploaded to the cluster after this.
+
+        Args:
+            job_config (JobConfig): The job config of driver.
+        """
+        pkg_uris = job_config.get_runtime_env_uris()
+        if len(pkg_uris) == 0:
+            return  # Return early to avoid internal kv check in this case.
+        for pkg_uri in pkg_uris:
+            if not package_exists(pkg_uri):
+                file_path = self._get_local_path(pkg_uri)
+                pkg_file = Path(file_path)
+                working_dir = job_config.runtime_env.get("working_dir")
+                py_modules = job_config.runtime_env.get("py_modules")
+                excludes = job_config.runtime_env.get("excludes") or []
+                _logger.info(
+                    f"{pkg_uri} doesn't exist. Create new package with"
+                    f" {working_dir} and {py_modules}")
+                if not pkg_file.exists():
+                    create_project_package(working_dir, py_modules, excludes,
+                                           file_path)
+                # Push the data to remote storage
+                pkg_size = push_package(pkg_uri, pkg_file)
+                _logger.info(
+                    f"{pkg_uri} has been pushed with {pkg_size} bytes")
+
+    def ensure_runtime_env_setup(self, pkg_uris: List[str]) -> Optional[str]:
+        """Make sure all required packages are downloaded it local.
+
+        Necessary packages required to run the job will be downloaded
+        into local file system if it doesn't exist.
+
+        Args:
+            pkg_uri list(str): Package of the working dir for the runtime env.
+
+        Return:
+            Working directory is returned if the pkg_uris is not empty,
+            otherwise, None is returned.
+        """
+        pkg_dir = None
+        for pkg_uri in pkg_uris:
+            # For each node, the package will only be downloaded one time
+            # Locking to avoid multiple process download concurrently
+            pkg_file = Path(self._get_local_path(pkg_uri))
+            with FileLock(str(pkg_file) + ".lock"):
+                pkg_dir = self.fetch_package(pkg_uri)
+            sys.path.insert(0, str(pkg_dir))
+
+        # Right now, multiple pkg_uris are not supported correctly.
+        # We return the last one as working directory
+        return str(pkg_dir) if pkg_dir else None
+
+    def delete_uri(self, uri: str) -> bool:
+        """Deletes a specific URI from the local filesystem.
+
+        Args:
+            uri (str): URI to delete.
+
+        Returns:
+            True if the URI was successfully deleted, else False.
+        """
+        deleted = False
+        path = Path(self._get_local_path(uri))
+        with FileLock(str(path) + ".lock"):
```

Comment:
What's the lock for here? Is it in case we have multiple WorkingDirManagers in different processes?

Reply:
We still have installation/deletion happening from multiple threads. Ideally the concurrency control would all happen in the agent, but that isn't quite implemented yet.